plugins/examples/script-expert-scoring/src/main/java/org/elasticsearch/example/expertscript/ExpertScriptPlugin.java

```diff
@@ -0,0 +1,158 @@
+                        PostingsEnum postings = context.reader().postings(new Term(field, term));
+                        if (postings == null) {
+                            // the field and/or term don't exist in this segment, so always return 0
+                            return () -> 0.0d;
+                        }
+                        return new LeafScript() {
+                            int currentDocid = -1;
+                            @Override
+                            public void setDocument(int docid) {
+                                // advance has undefined behavior calling with a docid <= its current docid
+                                if (postings.docID() < docid) {
+                                    try {
+                                        postings.advance(docid);
+                                    } catch (IOException e) {
+                                        throw new UncheckedIOException(e);
+                                    }
+                                }
+                                currentDocid = docid;
+                            }
+                            @Override
+                            public double runAsDouble() {
+                                if (postings.docID() != currentDocid) {
+                                    // advance moved past the current doc, so this doc has no occurrences of the term
+                                    return 0.0d;
+                                }
+                                try {
+                                    return postings.freq();
+                                } catch (IOException e) {
+                                    throw new UncheckedIOException(e);
+                                }
+                            }
+                        };
+                    }
+
+                    @Override
+                    public boolean needsScores() {
+                        return false;
+                    }
+                };
+            }
+            throw new IllegalArgumentException("Unknown script name " + scriptSource);
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public SearchScript search(CompiledScript compiledScript, SearchLookup lookup, @Nullable Map<String, Object> params) {
+          Function<Map<String,Object>,SearchScript> scriptFactory = (Function<Map<String,Object>,SearchScript>) compiledScript.compiled();
+          return scriptFactory.apply(params);
+        }
+
+        @Override
+        public ExecutableScript executable(CompiledScript compiledScript, @Nullable Map<String, Object> params) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean isInlineScriptEnabled() {
+            return true;
+        }
+
+        @Override
+        public void close() {}
+    }
+    // end::expert_engine
```

Comment:
I think this is a good enough tweak I will try to move these into LeafSearchScript.